### DISCUSSION TITLE
Set compiler options only for a custom/external module

### DISCUSSION THREAD

**Discussion by yguel** [zephyrproject-rtos/zephyr#52485] [General]
** Problem **
When defining a custom/external module, we need some specific compilation options:
1. removing some warnings (just for the module) example: -Wpointer-arith
2. setting specific preprocessor variables -DENABLE_X=1

** Expected solution **
The solution we would like would be to get the current compilation options from one or multiple variable and be able to modify it from the CMakeLists.txt of the module

** Tests done **
We try the following alternatives without success:
1. Setting CONFIG_COMPILER_OPT
2. Setting zephyr_library_compile_options in the CMakeLists.txt of the module
3. Setting target_compile_options(app PRIVATE ${MY_OPTIONS}) in the CMakeLists.txt of the module

like suggested in the following issues:
1. https://github.com/zephyrproject-rtos/zephyr/issues/15065
2. https://github.com/zephyrproject-rtos/zephyr/issues/41708

** Ideas **
We have seen that some modules include xxxx.cmake files in a cmake repository in the module directory.
We are wondering if this is the kind of approach we should take.

**Comment by nordicjm** [comment]
> Setting target_compile_options(app PRIVATE ${MY_OPTIONS}) in the CMakeLists.txt of the module

Why would changing the compile options of the app from a module do anything? You'd need to use the name of what you've configured the module as

**Comment by yguel** [comment]
@nordicjm we also try the name of the directory of the module instead of app, but it did not do anything.
Is there a special CMake/Zephyr command to set the name of the module ?

**Comment by nordicjm** [comment]
Moving to a discussion as this is not a bug.

A good starting example would be zcbor, the zcbor cmake files are stored internally in the zephyr tree here https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/zcbor/CMakeLists.txt but are used by the zcbor module, you don't need to follow this part just the part of how files are added.

If you add this line for example:
```
  zephyr_library()
  zephyr_library_sources(
    ${ZEPHYR_ZCBOR_MODULE_DIR}/src/zcbor_common.c
    ${ZEPHYR_ZCBOR_MODULE_DIR}/src/zcbor_decode.c
    ${ZEPHYR_ZCBOR_MODULE_DIR}/src/zcbor_encode.c
  )

+zephyr_compile_options(-Wdemo_fail)

  zephyr_compile_definitions_ifdef(CONFIG_ZCBOR_CANONICAL ZCBOR_CANONICAL)
```

And try to build a zephyr application that includes zcbor, you will see it fails:
```
FAILED: zephyr/subsys/mgmt/mcumgr/grp/img_mgmt/CMakeFiles/mgmt_mcumgr_grp_img.dir/src/img_mgmt.c.obj 
ccache /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -DNRF52840_XXAA -D__PROGRAM_START -D__ZEPHYR__=1 -I/tmp/aa/zephyr/subsys/mgmt/mcumgr/grp/img_mgmt/include -I/tmp/aa/zephyr/include -I/tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/_AA/zephyr/include/generated -I/tmp/aa/zephyr/soc/arm/nordic_nrf/nrf52 -I/tmp/aa/zephyr/soc/arm/nordic_nrf/common/. -I/tmp/aa/zephyr/subsys/mgmt/mcumgr/util/include -I/tmp/aa/modules/hal/cmsis/CMSIS/Core/Include -I/tmp/aa/modules/hal/nordic/nrfx -I/tmp/aa/modules/hal/nordic/nrfx/drivers/include -I/tmp/aa/modules/hal/nordic/nrfx/mdk -I/tmp/aa/zephyr/modules/hal_nordic/nrfx/. -I/tmp/aa/modules/debug/segger/SEGGER -I/tmp/aa/modules/debug/segger/Config -I/tmp/aa/zephyr/modules/segger/. -I/tmp/aa/modules/lib/zcbor/include -I/tmp/aa/zephyr/subsys/mgmt/mcumgr/mgmt/include -I/tmp/aa/bootloader/mcuboot/boot/bootutil/zephyr/.. -I/tmp/aa/bootloader/mcuboot/boot/bootutil/zephyr/../include -I/tmp/aa/bootloader/mcuboot/boot/bootutil/zephyr/../../zephyr/include -isystem /tmp/aa/zephyr/lib/libc/minimal/include -isystem /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/include -isystem /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/include-fixed -fno-strict-aliasing -Os -imacros /tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/_AA/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfp16-format=ieee --sysroot=/opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/arm-zephyr-eabi -imacros /tmp/aa/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr=CMAKE_SOURCE_DIR -fmacro-prefix-map=/tmp/aa/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/tmp/aa=WEST_TOPDIR -ffunction-sections -fdata-sections -Wdemo_fail -std=c99 -nostdinc -MD -MT zephyr/subsys/mgmt/mcumgr/grp/img_mgmt/CMakeFiles/mgmt_mcumgr_grp_img.dir/src/img_mgmt.c.obj -MF zephyr/subsys/mgmt/mcumgr/grp/img_mgmt/CMakeFiles/mgmt_mcumgr_grp_img.dir/src/img_mgmt.c.obj.d -o zephyr/subsys/mgmt/mcumgr/grp/img_mgmt/CMakeFiles/mgmt_mcumgr_grp_img.dir/src/img_mgmt.c.obj -c /tmp/aa/zephyr/subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt.c
arm-zephyr-eabi-gcc: error: unrecognized command-line option '-Wdemo_fail'
ninja: build stopped: subcommand failed.
```

Thus, it works.

**Reply by yguel** [reply]
Well it does not exactly work as we want since we need to remove some options. So it does not replace the actual options used but add -Wdemo_fail.
to be explicit: we want to remove the -Wpointer_arith only for our module, but 
if ${WANTED_OPTION} is ALL_CURRENT_OPTION - {-Wpointer_arith}
using
zephyr_compile_options(${WANTED_OPTION}) does nothing -Wpointer_arith is still in the list of options.

By the way, that is why I flagged my issue as enhancement (not bug).

**Reply by tejlmand** [reply]
> Well it does not exactly work as we want since we need to remove some options.

you don't need to remove options, just append with the `no-` variant.

gcc process the options in the order specified and the last variant wins, for example compiling this file:
```
// f.c
void f (void *p)
{
  p += sizeof(unsigned int);
}
```
now let's compile and see the warnings:
```
$ /opt/zephyr-sdk-0.15.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -Wpointer-arith -o f.o -c f.c
f.c: In function 'f':
f.c:4:5: warning: pointer of type 'void *' used in arithmetic [-Wpointer-arith]
    4 |   p += sizeof(unsigned int);
      |     ^~
```
and let's append the `no-` version of the flag.
```
$ /opt/zephyr-sdk-0.15.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -Wpointer-arith -Wno-pointer-arith -o f.o -c f.c
$
```
and the warning is gone.

This is the standard way in gcc to disable an already enabled flag.

You should be able to use `zephyr_library_compile_options()` or set the flag directly on the source file, like this:
```
set_source_files_properties(<file> PROPERTIES COMPILE_OPTIONS "-Wno-pointer-arith")
```
depending on your exact need.
Note, that if this results in you opening a PR, then most likely it will not be directly approved to set it like this, as we would prefer an entry in our compiler flags file due to variance between compilers, but downstream this should be a fully functional and simple solution.

**Reply by yguel** [reply]
Dear tejlmand,
Thanks for your detailed answer. It works.
Side note: it is a bit hard to isolate the code that produces the warning since it was from functions declared in headers included and used in many places in the project (it is not possible to set_source_files_properties for header).
It seems that the solution is to write interface functions that act as interface and to include the headers that causes the warning only in one place where the interface functions are defined. Our next step is to try this solution.

**Comment by xhpohanka** [comment]
I found this discussion searching for my issue.
Can I have additional question? I have an external module with its own `CMakeLists.txt`. Can I somehow add module options from applications `CMakeLists.txt`?
